sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -171,6 +171,13 @@ trait Partitioning {
     required.requiredNumPartitions.forall(_ == numPartitions) && satisfies0(required)
   }
 
+  final def changePartitionNumber(
```

Comment:
I'd make this an abstract method and ask all sub-classes to implement it. For partitioning that can't change num partitions, they should return `UnknownPartitioning`